### ISSUE TITLE
Ray sphere

### ISSUE THREAD

**Issue by jb15-creator** [jb15-creator/Python-code#2]
import pygame
import numpy as np
import sys
import math

# === INITIALIZE ===
pygame.init()
pygame.mixer.init(frequency=44100, size=-16, channels=2, buffer=512)

# === CONFIG (MOBILE-FRIENDLY) ===
SCREEN_WIDTH = 800
SCREEN_HEIGHT = 500
SCALE = 40
CAMERA_DISTANCE = 12.0
CAMERA_HEIGHT = 2.5
ANIMATION_SPEED_MULTIPLIER = 25.0
NUM_RAYS = 60
MAX_BOUNCES = 4

# === SOUND / PHYSICS SETTINGS ===
SAMPLE_RATE = 44100
FREQUENCY = 220
DURATION = 2.0
CUTOFF_FREQ = 1000
RC = 1.0 / (2.0 * np.pi * CUTOFF_FREQ)
dt = 1.0 / SAMPLE_RATE
alpha = dt / (RC + dt)

# ADSR envelope
ATTACK_TIME = 0.05
DECAY_TIME = 0.1
SUSTAIN_LEVEL = 0.7
RELEASE_TIME = 0.3

# Room / source / receiver setup
ROOM_SIZE = np.array([8.0, 6.0, 5.0])
SOUND_SPEED = 343.0
SOURCE_POS = np.array([2.0, 3.0, 2.5])
RECEIVER_POS = np.array([6.0, 3.0, 2.5])
RECEIVER_RADIUS = 0.3

# === COLORS ===
WHITE = (255, 255, 255)
BLACK = (0, 0, 0)
BLUE = (70, 130, 180)
DARK_BLUE = (50, 90, 140)
RED = (200, 50, 50)
GREEN = (50, 200, 50)
YELLOW = (255, 255, 100)
GRAY = (150, 150, 150)
CYAN = (100, 200, 200)

# === FUNCTIONS ===

def apply_adsr(signal, sample_rate, attack, decay, sustain_level, release):
    num_samples = len(signal)
    envelope = np.ones(num_samples)
    a = int(attack * sample_rate)
    d = int(decay * sample_rate)
    r = int(release * sample_rate)
    if a > 0:
        envelope[:a] = np.linspace(0, 1, a)
    if d > 0:
        envelope[a:a+d] = np.linspace(1, sustain_level, d)
    if r > 0:
        envelope[-r:] = np.linspace(sustain_level, 0, r)
    envelope[a+d:-r] = sustain_level
    return signal * envelope

def generate_saw_wave(frequency, duration, sample_rate):
    num_samples = int(duration * sample_rate)
    t = np.linspace(0, duration, num_samples, False)
    saw_wave = 2 * (t * frequency - np.floor(t * frequency + 0.5))
    return saw_wave

def apply_rc_lowpass(signal, alpha):
    filtered = np.zeros_like(signal)
    filtered[0] = signal[0]
    for i in range(1, len(signal)):
        filtered[i] = alpha * signal[i] + (1 - alpha) * filtered[i - 1]
    return filtered

def create_sound():
    saw = generate_saw_wave(FREQUENCY, DURATION, SAMPLE_RATE)
    filtered = apply_rc_lowpass(saw, alpha)
    enveloped = apply_adsr(filtered, SAMPLE_RATE, ATTACK_TIME, DECAY_TIME, SUSTAIN_LEVEL, RELEASE_TIME)
    audio = (enveloped * 32767).astype(np.int16)
    stereo = np.column_stack((audio, audio))
    return pygame.sndarray.make_sound(stereo)

def rotate_point(point, angle_x, angle_y):
    x = point[0] * np.cos(angle_y) - point[2] * np.sin(angle_y)
    z = point[0] * np.sin(angle_y) + point[2] * np.cos(angle_y)
    y = point[1]
    y2 = y * np.cos(angle_x) - z * np.sin(angle_x)
    z2 = y * np.sin(angle_x) + z * np.cos(angle_x)
    return np.array([x, y2, z2])

def project_3d_to_2d(point_3d, ax, ay):
    centered = point_3d - ROOM_SIZE / 2
    rotated = rotate_point(centered, ax, ay)
    x, y, z = rotated
    z += CAMERA_DISTANCE
    if z <= 0.1:
        z = 0.1
    factor = SCALE / z * 10
    return (int(SCREEN_WIDTH / 2 + x * factor),
            int(SCREEN_HEIGHT / 2 - y * factor), z)

# === SURFACE-DEPENDENT REFLECTION FUNCTION ===
def reflect_ray(position, direction, room_size):
    """
    Reflect ray with surface-dependent diffusion and reflectivity.
    Returns (new_position, new_direction, reflected_flag, energy_loss).
    """

    new_pos = position.copy()
    new_dir = direction.copy()
    reflected = False

    # Material properties (diffuse_strength, reflectivity_loss)
    wall_props = {
        "x_min": (0.15, 0.8),
        "x_max": (0.15, 0.8),
        "y_min": (0.05, 0.9),
        "y_max": (0.05, 0.9),
        "z_min": (0.10, 0.85),
        "z_max": (0.10, 0.85),
    }

    for axis in range(3):
        if new_pos[axis] < 0:
            new_pos[axis] = 0
            new_dir[axis] *= -1
            key = ["x_min", "y_min", "z_min"][axis]
            diffuse_strength, reflectivity = wall_props[key]
            reflected = True
        elif new_pos[axis] > room_size[axis]:
            new_pos[axis] = room_size[axis]
            new_dir[axis] *= -1
            key = ["x_max", "y_max", "z_max"][axis]
            diffuse_strength, reflectivity = wall_props[key]
            reflected = True

        if reflected:
            new_dir += np.random.normal(scale=diffuse_strength, size=3)
            norm = np.linalg.norm(new_dir)
            if norm > 0:
                new_dir /= norm
            return new_pos, new_dir, reflected, reflectivity

    return new_pos, new_dir, reflected, 1.0

class Ray3D:
    def __init__(self, start, direction):
        self.path = [start.copy()]
        self.current = start.copy()
        self.dir = direction / np.linalg.norm(direction)
        self.bounces = 0
        self.active = True
        self.hit_receiver = False
        self.hit_time = 0
        self.amp = 1.0
        self.dist = 0

    def update(self, dt, room, recv_pos, recv_r):
        if not self.active:
            return
        move = SOUND_SPEED * ANIMATION_SPEED_MULTIPLIER * dt
        new_pos = self.current + self.dir * move

        # Handle reflections
        reflected_pos, reflected_dir, did_reflect, energy_loss = reflect_ray(new_pos, self.dir, room)
        if did_reflect:
            self.bounces += 1
            self.amp *= energy_loss
            self.path.append(reflected_pos.copy())
            self.dir = reflected_dir
            if self.bounces >= MAX_BOUNCES:
                self.active = False
                return

        self.dist += np.linalg.norm(reflected_pos - self.current)
        self.current = reflected_pos

        # Check receiver hit
        if not self.hit_receiver:
            if np.linalg.norm(self.current - recv_pos) < recv_r:
                self.hit_receiver = True
                self.hit_time = self.dist / SOUND_SPEED
                self.amp *= 1.0 / (1.0 + self.dist * 0.05)
                self.active = False

def draw_3d_box(screen, size, ax, ay):
    c = [np.array([0, 0, 0]),
         np.array([size[0], 0, 0]),
         np.array([size[0], size[1], 0]),
         np.array([0, size[1], 0]),
         np.array([0, 0, size[2]]),
         np.array([size[0], 0, size[2]]),
         np.array([size[0], size[1], size[2]]),
         np.array([0, size[1], size[2]])]
    edges = [(0,1),(1,2),(2,3),(3,0),(4,5),(5,6),(6,7),(7,4),(0,4),(1,5),(2,6),(3,7)]
    proj = [project_3d_to_2d(p, ax, ay) for p in c]
    for i, j in edges:
        if proj[i][2] > 0 and proj[j][2] > 0:
            pygame.draw.line(screen, GRAY, proj[i][:2], proj[j][:2], 2)

def draw_3d_sphere(screen, center, r, ax, ay, color):
    num = 16
    points = []
    for i in range(num):
        theta = 2 * math.pi * i / num
        p = center + np.array([r * math.cos(theta), 0, r * math.sin(theta)])
        proj = project_3d_to_2d(p, ax, ay)
        points.append(proj[:2])
    pygame.draw.lines(screen, color, True, points, 2)

def draw_expanding_sphere(screen, center, r, ax, ay):
    if r < 0.1:
        return
    num = 32
    pts = []
    for i in range(num):
        theta = 2 * math.pi * i / num
        p = center + np.array([r * math.cos(theta), 0, r * math.sin(theta)])
        proj = project_3d_to_2d(p, ax, ay)
        if proj[2] > 0:
            pts.append(proj[:2])
    if len(pts) > 2:
        pygame.draw.lines(screen, YELLOW, True, pts, 1)

def draw_3d_ray(screen, ray, ax, ay):
    pts = [project_3d_to_2d(p, ax, ay)[:2] for p in ray.path]
    pts.append(project_3d_to_2d(ray.current, ax, ay)[:2])
    if len(pts) > 1:
        color = GREEN if ray.hit_receiver else CYAN
        pygame.draw.lines(screen, color, False, pts, 1)

def create_fibonacci_sphere_directions(n):
    dirs = []
    phi = np.pi * (3. - np.sqrt(5.))
    for i in range(n):
        y = 1 - (i / float(n - 1)) * 2
        r = np.sqrt(1 - y*y)
        theta = phi * i
        x = np.cos(theta) * r
        z = np.sin(theta) * r
        dirs.append(np.array([x, y, z]))
    return dirs

# === SETUP ===
screen = pygame.display.set_mode((SCREEN_WIDTH, SCREEN_HEIGHT))
pygame.display.set_caption("3D Sound Reflection Simulation")
sound = create_sound()
clock = pygame.time.Clock()
running = True
playing = False
animation_time = 0
rays = []
received = []
radius = 0
ax, ay = 0.3, 0.5

# === BUTTON ===
BUTTON_WIDTH = 200
BUTTON_HEIGHT = 60
button_rect = pygame.Rect(
    SCREEN_WIDTH // 2 - BUTTON_WIDTH // 2,
    SCREEN_HEIGHT - BUTTON_HEIGHT - 20,
    BUTTON_WIDTH,
    BUTTON_HEIGHT
)

# === MAIN LOOP ===
while running:
    dt = clock.tick(60) / 1000
    for e in pygame.event.get():
        if e.type == pygame.QUIT:
            running = False
        elif e.type == pygame.MOUSEBUTTONDOWN:
            if button_rect.collidepoint(e.pos):
                sound.play()
                playing = True
                animation_time = 0
                radius = 0
                received = []
                dirs = create_fibonacci_sphere_directions(NUM_RAYS)
                rays = [Ray3D(SOURCE_POS, d) for d in dirs]

    if playing:
        animation_time += dt
        radius = SOUND_SPEED * ANIMATION_SPEED_MULTIPLIER * animation_time
        for ray in rays:
            ray.update(dt, ROOM_SIZE, RECEIVER_POS, RECEIVER_RADIUS)
            if ray.hit_receiver and ray not in received:
                received.append(ray)
        if all(not r.active for r in rays) and animation_time > 2.0:
            playing = False

    screen.fill(BLACK)
    draw_3d_box(screen, ROOM_SIZE, ax, ay)
    if playing and radius < 15:
        draw_expanding_sphere(screen, SOURCE_POS, radius, ax, ay)
    for ray in rays:
        draw_3d_ray(screen, ray, ax, ay)
    draw_3d_sphere(screen, SOURCE_POS, 0.2, ax, ay, RED)
    draw_3d_sphere(screen, RECEIVER_POS, RECEIVER_RADIUS, ax, ay, GREEN)

    # BUTTON
    btn_color = DARK_BLUE if playing else BLUE
    pygame.draw.rect(screen, btn_color, button_rect, border_radius=12)
    pygame.draw.rect(screen, WHITE, button_rect, 2, border_radius=12)
    font = pygame.font.Font(None, 36)
    text = font.render("Play Sound", True, WHITE)
    screen.blit(text, text.get_rect(center=button_rect.center))

    pygame.display.flip()

pygame.quit()
sys.exit()
